### PR TITLE
Add ccp cleared orders

### PR DESCRIPTION
**Pull request by alexmatson-da**

### CODE REVIEW

File: exberry_adapter/bot/exberry_adapter_bot.py
Comment (alexmatson-da) [inline exberry_adapter/bot/exberry_adapter_bot.py:133]: I know this is a pretty clunky pattern...

Comment (alexmatson-da) [reply exberry_adapter/bot/exberry_adapter_bot.py]: Turns out this was not only clunky, but buggy... changed the approach

File: daml/Marketplace/ExchangeParticipant.daml
Comment (alexmatson-da) [inline daml/Marketplace/ExchangeParticipant.daml:166]: Where should the CCP party come from? For now I just made it a parameter of the choice, and on the UI it's currently hardcoded to `"Ccp"` when the choice is exercised. 

If we make it a dropdown on the UI that lists the CCPs that the investor is a customer of, then different investors might enter different CCPs for the same (cleared) market pair, which seems... wrong

Comment (alexgraham-da) [reply daml/Marketplace/ExchangeParticipant.daml:166]: Hmmm that's a good point. I'll have to ask Eric about customers having multiple CCPs, but I'm guessing that each cleared market will have one CCP attached to it. 

Come to think of it, the CCP already needs to know who the exchange is in order to request `FairValue`s, so I think we'll have to add a specific CCP <-> Exchange relationship.

File: daml/Marketplace/ExchangeParticipant.daml
Comment (alexmatson-da) [inline daml/Marketplace/ExchangeParticipant.daml:195]: I don't know if this is "elegant" or "hacky", but I made sure that an investor placing an order in a cleared market has the authorization of the CCP by moving the `MakeClearedOrder` choice to its `CCPCustomer` contract. 

if the investor is not a customer, they won't have such a contract, and the trade placement is blocked by the key not found error!

Comment (alexgraham-da) [reply daml/Marketplace/ExchangeParticipant.daml:195]: This works, you could also use `lookupByKey` which returns an `Optional` to have a clearer abort message (and maybe keep it in the `ExchangeParticipant` function):

```
optCCPCustomer <- lookupByKey @CCPCustomer etc. etc.
assertMsg "Participant must be a customer etc etc" $ isSome optCCPCustomer
```

File: daml/Marketplace/Exchange.daml
Comment (alexmatson-da) [inline daml/Marketplace/Exchange.daml:142]: my haskell-foo is not strong enough to write a function that takes a shorter input string and a max length, and returns a padded string with as many instances of `0` to the end of the input string to fill up to the max length...

Comment (alexgraham-da) [reply daml/Marketplace/Exchange.daml:142]: may be a nicer way but:
```
addZeros : Int -> Text -> Text
addZeros n txt = T.implode $ T.explode txt <> replicate (n - T.length txt) "0"
```

Comment (alexmatson-da) [reply daml/Marketplace/Exchange.daml:142]: Gonna leave this as a simple "CLR" appending for now, and I'll look into creating unique instrument IDs this week / next week

File: ui/src/components/Investor/Investor.tsx
Comment (alexgraham-da) [inline ui/src/components/Investor/Investor.tsx:139]: Can probably take this out

File: ui/src/components/Exchange/MarketPairs.tsx
Comment (alexgraham-da) [inline ui/src/components/Exchange/MarketPairs.tsx:44]: Bob Loblaw's Law Blog

Comment (alexmatson-da) [reply ui/src/components/Exchange/MarketPairs.tsx:44]: I was very disappointed that I had to do this, typescript can't infer the fact that this is a cleared market so the first element of the outer tuple here will always be another tuple (the Id pair)

This error would (should) never get thrown, since it's just here to satisfy the compiler... but I will change the error message though 😆

File: ui/src/components/Investor/Investor.tsx
Comment (alexmatson-da) [reply ui/src/components/Investor/Investor.tsx:139]: Oops yeah, thought I got all of these. Thanks for catching!

Comment (alexmatson-da) [reply ui/src/components/Investor/Investor.tsx]: Found a few more unnecessary console.logs too, removed them

File: ui/src/components/common/TradeCard.tsx
Comment (alexmatson-da) [inline ui/src/components/common/TradeCard.tsx:29]: as a tidbit, the ClearedTrade also has a `trade.timestamp` field like the `SettledTradeSide`, but it was in a different format (an ISO string) rather than a UNIX timestamp, so the parsing resulted in the wrong time being displayed. 

The `eventTimestamp` was as expected, but we may wish to normalize this. And I'm also not sure why we have two different timestamp fields to begin with (this was present in DerivativeTradeSide as well I believe)

File: daml/Marketplace/ExchangeParticipant.daml
Comment (alexmatson-da) [reply daml/Marketplace/ExchangeParticipant.daml:196]: For some reason, `lookupByKey` wasn't working even when the contract did exist. I forgot what the error was exactly (tried it a week ago), but some detail about disclosure and contract privacy I think. 

I put some frontend validation in place, so if this situation happens we stop it before exercising, and we still avoid a random DAML LF error dump on the screen.

File: daml/Marketplace/Trading.daml
Comment (alexgraham-da) [inline daml/Marketplace/Trading.daml:292]: Please change these to `assertMsg`, it can get really tricky to debug with just `assert` since they give almost zero info.

File: ui/src/components/common/TradeCard.tsx
Comment (alexgraham-da) [reply ui/src/components/common/TradeCard.tsx:29]: In `ClearedTradeSide`, `eventTimestamp` is what we receive from Exberry, and `timestamp` is a Daml `getTime` for when the trade is novated. I believe in regular `TradeSide`, `timestamp` is the `eventTimestamp` from Exberry.

So I'd say it would be good to make this more consistent, either by having the `timestamp` field in regular trades be `eventTimestamp`, or making the `timestamp` in `ClearedTradeSide` be `timeNovated` or something to that effect. Or both.

**Review by alexgraham-da — APPROVED** [review]
This looks good! Just make sure you pull in `master` (since there were minor changes to `add-ccp` after PR) and change the destination to `master` before merging.

File: ui/src/components/common/TradeCard.tsx
Comment (alexgraham-da) [reply ui/src/components/common/TradeCard.tsx:29]: It might also be worth converting the `eventTimestamp` from Exberry into a Daml `Time` when creating the contracts in the adapter.

File: ui/src/components/Exchange/MarketPairs.tsx
Comment (alexgraham-da) [reply ui/src/components/Exchange/MarketPairs.tsx]: Weird that it's totally fine with you assuming it's a tuple only provided it isn't a string...what if it's literally any other type??

Comment (alexmatson-da) [reply ui/src/components/Exchange/MarketPairs.tsx]: That's the cool thing, the compiler knows at least it can't be any other type. On the Daml side, `clearedMarkets` contains items of type `((Id, Id), Party)`

and with the typescript codegen, this translates to `Tuple2<Tuple2<Id, Id>, String>` 

Then when calling `unwrapDamlTuple(listing)`, the type is converted to `(string | Tuple2<Id,Id>)[]` - an array of items that could either be strings or tuples, including the first item (although we know in advance it maps to the tuple, this may not be true just by the type) 

So in unwrapping, we actually lose some type information that we had before! Interesting, isn't it?

Comment (alexgraham-da) [reply ui/src/components/Exchange/MarketPairs.tsx]: Ohhhhh I see! Super interesting, thank you for the explanation

**Pull request merged**
